lib/models/cycles-metadata.ts

```diff
@@ -0,0 +1,111 @@
+import { BaseCyclesMetadata, PlainCyclesMetadata } from '../types/cycles-metadata';
+import {
+  ProperCycle,
+  PSALTER_WEEKS,
+  PsalterWeekCycle,
+  SUNDAY_CYCLES,
+  SundayCycle,
+  WEEKDAY_CYCLES,
+  WeekdayCycle,
+} from '../constants/cycles';
+import { RomcalConfig } from './config';
+import { getUtcDateFromString } from '../utils/dates';
+import { RomcalCalendarMetadata } from '../types/liturgical-day';
+
+/**
+ * A cycle metadata object, based on the contextual liturgical day & year
+ */
+export class CyclesMetadata implements BaseCyclesMetadata {
+  properCycle: ProperCycle;
+  #properCycleName?: string;
+  sundayCycle: SundayCycle;
+  #sundayCycleName?: string;
+  weekdayCycle: WeekdayCycle;
+  #weekdayCycleName?: string;
+  psalterWeek: PsalterWeekCycle;
+  #psalterWeekName?: string;
+  #config: RomcalConfig;
+
+  constructor(date: Date, calendar: RomcalCalendarMetadata, properCycle: ProperCycle, config: RomcalConfig) {
+    this.#config = config;
+
+    const year = parseInt(calendar.startOfLiturgicalYear, 10);
+
+    // Compute cycle of the liturgical year,
+    // and cache the data since they are the same for every day of the year
+    if (!this.#config.cyclesCache[year]) {
+      const firstSundayOfAdvent = getUtcDateFromString(calendar.startOfLiturgicalYear);
+
+      let sundayCycle: SundayCycle;
+      let weekdayCycle: WeekdayCycle;
+
+      // Formula to calculate Sunday cycle (Year A, B, C)
+      const thisSundayCycleIndex: number = (year - 1963) % 3;
```

Comment:
This make romcal (sort of) backward incompatible with older romcal versions and makes romcal generate calendars for years after `1963` (inclusively). That is not a big deal, but if one tried to generate calendar, e.g., for `1900`, I’d presume it _should_ fail calendar generation, but actually it does not fail and Easter Sunday is on 15 April 1900 (based on [this website](https://tlarsen2.tripod.com/thomaslarsen/easterdates.html), it is correct).

Still, I suggest to remove `1963` constant, as it is not required to calculate the Sunday cycle letter: simply use `year % 3` instead compare the remainder:
- `1` → `A`;
- `2` → `B`;
- `0` → `C`.

Reply:
This line of code is coming from older romcal version (at least 1.x), and could be Indeed improved.

Reply:
I think it should be changed (maybe in a separate PR), as I think that the historicality of the calendar should be set separately (e.g. using a romcal option, see #23).

Reply:
Fixed.